staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -929,3 +935,69 @@ func isWatchErrorRetriable(err error) bool {
 	}
 	return false
 }
+
+type initialEventsEndBookmarkTicker struct {
+	clock.Ticker
+	clock clock.Clock
+	name  string
+
+	watchStart           time.Time
+	tickInterval         time.Duration
+	lastEventObserveTime time.Time
+}
+
+func newInitialEventsEndBookmarkTicker(name string, c clock.Clock, watchStart time.Time, tickInterval time.Duration, exitOnInitialEventsEndBookmark *bool) *initialEventsEndBookmarkTicker {
+	exitOnInitialEventsEndBookmarkRequested := exitOnInitialEventsEndBookmark != nil
+	clockWithTicker, ok := c.(clock.WithTicker)
+	if !ok || !exitOnInitialEventsEndBookmarkRequested {
+		if exitOnInitialEventsEndBookmarkRequested {
+			klog.Warningf("clock does not support WithTicker interface but exitOnInitialEventsEndBookmark was requested")
+		}
+		return &initialEventsEndBookmarkTicker{
+			Ticker: &noopTicker{},
+		}
+	}
+
+	return &initialEventsEndBookmarkTicker{
+		Ticker:       clockWithTicker.NewTicker(tickInterval),
+		clock:        c,
+		name:         name,
+		watchStart:   watchStart,
+		tickInterval: tickInterval,
+	}
+}
+
+func (t *initialEventsEndBookmarkTicker) observeLastEventTimeStamp(lastEventObserveTime time.Time) {
+	t.lastEventObserveTime = lastEventObserveTime
+}
+
+func (t *initialEventsEndBookmarkTicker) warnIfExpired() {
+	if warningMsg := t.produceWarningIfExpired(); len(warningMsg) > 0 {
+		klog.Warning(warningMsg)
+	}
+}
+
+func (t *initialEventsEndBookmarkTicker) produceWarningIfExpired() string {
+	if _, ok := t.Ticker.(*noopTicker); ok {
+		return "" /*noop ticker*/
+	}
+	if t.lastEventObserveTime.IsZero() {
+		return fmt.Sprintf("%s: awaiting required bookmark event for initial events stream, no events received for %v", t.name, t.clock.Since(t.watchStart))
+	}
+	elapsedTime := t.clock.Now().Sub(t.lastEventObserveTime)
+	hasBookmarkTimerExpired := elapsedTime >= t.tickInterval
+
+	if !hasBookmarkTimerExpired {
+		return ""
```

Comment:
shouldn't we also Stop the ticker here?

The potential problem there is that I'm not sure we can call Stop() more than once on arbitrary ticker.
But I would like to avoid doing this processing indefinitely when we're already synchronized.

Reply:
> shouldn't we also Stop the ticker here?

no, the ticker ticks every `X` seconds. We are expired when the last time we observed an event is longer than the ticker interval. 

> But I would like to avoid doing this processing indefinitely when we're already synchronized.

as soon as we are synchronized we exit the `watchHandler` and stop the ticker. 
if streaming was not requested the ticker is a `noop` ticker.

Reply:
> as soon as we are synchronized we exit the watchHandler and stop the ticker.
if streaming was not requested the ticker is a noop ticker.

OK - that's right. Can you add a comment about that - it's important bit.

Reply:
sure, done.